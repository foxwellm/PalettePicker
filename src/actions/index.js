export const setCurrentPalette = (palette) => ({
  type: 'SET_CURRENT_PALETTE',
  palette
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
});

export const removeProject = (projectId) => ({
  type: 'REMOVE_PROJECT',
  projectId
});

export const setPalettes = (palettes) => ({
  type: 'SET_PALETTES',
  palettes
});

export const addPalette = (palette) => ({
  type: 'ADD_PALETTE',
  palette
});

export const removePalette = (paletteId) => ({
  type: 'REMOVE_PALETTE',
  paletteId
});

export const setSearchProjects = (search) => ({
  type: 'SEARCH_PROJECTS',
  search
});

export const setSearching = (bool) => ({
  type: 'SEARCHING_PROJECTS',
  bool
});

export const setLoading = (loading) => ({
  type: 'SET_LOADING',
  loading
});

export const setError = (error) => ({
  type: 'SET_ERROR',
  error
});