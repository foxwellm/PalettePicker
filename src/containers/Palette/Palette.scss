.palette-div {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(200px, 200px));
  grid-gap: 10px;
  justify-content: center;
}

.color-div {
  height: 300px;
  width: 200px;
  border: 1px solid black;
}

.generate-button {
  display: block;
  margin: 20px auto 40px auto;
}

.lock-button {
  height: 25px;
  width: 25px;
  border: none;
  border-radius: none;
  box-shadow: none;
  margin: 200px auto 10px auto;
}

.lock {
  display: block;
  background: url('./images/padlock.png');
  background-size: cover;
}

.unlock {
  display: block;
  background: url('./images/padlock-unlock.png');
  background-size: cover;
}

h4 {
  text-align: center;
}