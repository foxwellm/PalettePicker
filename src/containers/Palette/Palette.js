import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setCurrentPalette } from '../../actions';

export class Palette extends Component {
  constructor(props) {
    super(props);
    this.state = {
      locked: []
    }
  }
  
  setRandomColors = () => {
    const { locked } = this.state;
    const { currentPalette } = this.props;
    let newColors = [];
    for (let i = 0; i < 5; i++) {
      const color = locked.includes(i.toString()) ? currentPalette[i] : `#${Math.random().toString(16).slice(2, 8)}`;
      newColors.push(color.toUpperCase());
    }
    this.props.setCurrentPalette(newColors);
  }

  toggleLock = (e) => {
    const { locked } = this.state;
    const id = e.target.id;
    const newLocked = locked.includes(id) ? locked.filter(lockedId => lockedId !== id) : [...locked, id];
    this.setState({ locked: newLocked });
  }

  componentDidMount = () => {
    this.setRandomColors();
  }

  render() {
    const { locked } = this.state;
    const { currentPalette } = this.props;
    return(
      <div>
        <div className='palette-div'>
          {
            currentPalette.map((color, i) => {
              return(
                <div style={{ backgroundColor: color }} className='color-div'>
                  <button 
                    onClick={this.toggleLock}
                    className={locked.includes(i.toString()) ? 'lock lock-button' : 'unlock lock-button'}
                    id={i}
                  >
                  </button>
                  <h4>{color}</h4>
                </div>
              )
            })
          }
        </div>
        <button onClick={this.setRandomColors} className='generate-button standard-button'>Generate Palette</button>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  currentPalette: state.currentPalette,
});

const mapDispatchToProps = (dispatch) => ({
  setCurrentPalette: (palette) => dispatch(setCurrentPalette(palette)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Palette);