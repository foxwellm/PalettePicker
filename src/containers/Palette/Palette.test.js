import React from 'react';
import { Palette, mapStateToProps, mapDispatchToProps } from './Palette';
import { shallow } from 'enzyme';
import { setCurrentPalette } from '../../actions';

describe('Palette', () => {
  let wrapper;
  let mockCurrentPalette;
  let mockSetCurrentPalette;
  let mockEvent;

  beforeEach(() => {
    mockSetCurrentPalette = jest.fn();
    mockCurrentPalette = ['#ffffff', '#ffffff', '#ffffff', '#ffffff', '#ffffff'];
    mockEvent = { target: { id: '1' } };

    wrapper = shallow(
      <Palette
        currentPalette={mockCurrentPalette}
        setCurrentPalette={mockSetCurrentPalette}
      />
    )
  });

  it('should match the correct snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state()).toEqual({ locked: [] });
  });

  describe('componentDidMount', () => {
    it('should call setRandomColors', () => {
      wrapper.instance().setRandomColors = jest.fn();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().setRandomColors).toHaveBeenCalled();
    });
  });

  describe('setRandomColors', () => {
    
  });

  describe('toggleLock', () => {
    
  });

  describe('mapStateToProps', () => {
    it('should return an object of props', () => {
      const mockState = {
        currentPalette: mockCurrentPalette,
        otherState: false,
      };
      const expected = {
        currentPalette: mockCurrentPalette,
      };
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when setCurrentPalette is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCurrentPalette();
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentPalette();
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});