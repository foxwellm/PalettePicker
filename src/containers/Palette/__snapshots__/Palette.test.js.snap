// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Palette should match the correct snapshot 1`] = `
<div>
  <div
    className="palette-div"
  >
    <div
      className="color-div"
      key="0"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    >
      <button
        className="unlock lock-button"
        id={0}
        onClick={[Function]}
      />
      <h4>
        #FFFFFF
      </h4>
    </div>
    <div
      className="color-div"
      key="1"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    >
      <button
        className="unlock lock-button"
        id={1}
        onClick={[Function]}
      />
      <h4>
        #FFFFFF
      </h4>
    </div>
    <div
      className="color-div"
      key="2"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    >
      <button
        className="unlock lock-button"
        id={2}
        onClick={[Function]}
      />
      <h4>
        #FFFFFF
      </h4>
    </div>
    <div
      className="color-div"
      key="3"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    >
      <button
        className="unlock lock-button"
        id={3}
        onClick={[Function]}
      />
      <h4>
        #FFFFFF
      </h4>
    </div>
    <div
      className="color-div"
      key="4"
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    >
      <button
        className="unlock lock-button"
        id={4}
        onClick={[Function]}
      />
      <h4>
        #FFFFFF
      </h4>
    </div>
  </div>
  <button
    className="generate-button standard-button"
    onClick={[Function]}
  >
    Generate Palette
  </button>
</div>
`;
